src/scripts/shards/Shards.ts

```diff
@@ -70,6 +88,23 @@ class Shards implements Feature {
             this.shardUpgrades[typeNum * Shards.nEffects + effectNum] ++;
         }
     }
+    
+    private refundShardUpgrade(
+        typeNum: PokemonType,
+        effectNum: GameConstants.TypeEffectiveness
+    ) {
+        while (this.shardUpgrades[typeNum * Shards.nEffects + effectNum] > 0) {
+            this.shardUpgrades[typeNum * Shards.nEffects + effectNum]--;
+            this.gainShards(this.getShardUpgradeCost(typeNum, effectNum), typeNum);
+        }
+    }
```

Comment:
I don't think we need to have this as a method of Shards, as it won't be used other than for updating a game save.